core/types/transaction.go

```diff
@@ -128,150 +89,140 @@ func isProtectedV(V *big.Int) bool {
 	return true
 }
 
-// EncodeRLP implements rlp.Encoder
 func (tx *Transaction) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, &tx.data)
+	var l *LegacyTransaction
+	l = tx.inner.(*LegacyTransaction)
+	return rlp.Encode(w, l)
 }
 
-// DecodeRLP implements rlp.Decoder
 func (tx *Transaction) DecodeRLP(s *rlp.Stream) error {
 	_, size, _ := s.Kind()
-	err := s.Decode(&tx.data)
+	var l *LegacyTransaction
+	err := s.Decode(&l)
 	if err == nil {
 		tx.size.Store(common.StorageSize(rlp.ListSize(size)))
 		tx.time = time.Now()
 	}
+	tx.inner = l
 	return err
 }
 
 // MarshalJSON encodes the web3 RPC transaction format.
 func (tx *Transaction) MarshalJSON() ([]byte, error) {
 	hash := tx.Hash()
-	data := tx.data
-	data.Hash = &hash
-	return data.MarshalJSON()
+	return tx.inner.MarshalJSONWithHash(&hash)
 }
 
 // UnmarshalJSON decodes the web3 RPC transaction format.
 func (tx *Transaction) UnmarshalJSON(input []byte) error {
-	var dec txdata
-	if err := dec.UnmarshalJSON(input); err != nil {
+	type id struct {
+		Type *hexutil.Uint64 `json:"type" rlp:"-"`
+		Hash *common.Hash    `json:"hash" rlp:"-"`
+	}
+
+	var dec id
+	if err := json.Unmarshal(input, &dec); err != nil {
```

Comment:
Any thoughts on hoisting the hash out the `inner` tx type? The alternative is to put `Hash` in all `inner` tx types